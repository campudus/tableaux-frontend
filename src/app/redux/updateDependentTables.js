// @flow

import f from "lodash/fp";
import { set } from "lodash"; // eslint-disable-line

import { ColumnKinds } from "../constants/TableauxConstants";
import {
  doto,
  mapIndexed,
  mapPromise,
  memoizeWith,
  when
} from "../helpers/functools";
import { makeRequest } from "../helpers/apiHelper";
import getDisplayValue from "../helpers/getDisplayValue";
import route from "../helpers/apiRoutes.js";

export const calcColumnDependencies = columnCollection => {
  const filterLinkColumns = ([tableId, { data }]) => [
    f.parseInt(10, tableId),
    f.filter(isLinkColumn, data)
  ];

  const extractToTables = ([tableId, data]) =>
    f.map(({ id, toTable }) => ({ tableId, toTable, columnId: id }), data);

  const toDependencyMap = f.reduce((accum, { tableId, toTable, columnId }) => {
    accum[toTable] = accum[toTable] || {};
    const existingColumns = accum[toTable][tableId];
    accum[toTable][tableId] = existingColumns
      ? [...existingColumns, columnId]
      : [columnId];
    return accum;
  }, {});

  return doto(
    columnCollection,
    f.toPairs,
    f.map(filterLinkColumns),
    f.flatMap(extractToTables),
    toDependencyMap
  );
};

const listOfTableIds = f.compose(
  f.join(","),
  f.sortBy(f.identity),
  f.keys
);

const getCachedDependencyMap = memoizeWith(
  listOfTableIds,
  calcColumnDependencies
);

const isLinkColumn = f.propEq("kind", ColumnKinds.link);

export const performRowDeletion = (tableId, rowId, state) => {
  const dependencies = getCachedDependencyMap(state.columns);

  const removeRowFromState = f.update(
    ["rows", tableId, "data"],
    f.remove(f.propEq("id", rowId))
  );

  const hasDependants = () => !f.isEmpty(dependencies[tableId]);

  return doto(
    state,
    removeRowFromState,
    when(hasDependants, propagateRowDelete(tableId, rowId)),
    f.prop("rows")
  );
};

export const propagateRowDelete = f.curryN(
  3,
  (originTableId, originRowId, state) => {
    const updatedRows = state.rows;
    const removeLinkFrom = f.remove(f.propEq("id", originRowId));

    // in-place update
    const tableIds = f.keys(state.rows);

    tableIds.forEach(tableId => {
      if (tableId === originTableId) return;
      const columns = state.columns[tableId].data;

      columns.forEach(({ kind, toTable }, idx) => {
        const rows = state.rows[tableId].data;

        if (kind === ColumnKinds.link && toTable === originTableId) {
          rows.forEach((row, rowIdx) => {
            updatedRows[tableId].data[rowIdx].values[idx] = removeLinkFrom(
              state.rows[tableId].data[rowIdx].values[idx]
            );
          });
        }
      });
    });
    return { rows: updatedRows };
  }
);

export const hasTransitiveDependencies = (tableId, state) => {
  const dependencies = getCachedDependencyMap(state);
  const primaryDependencies = doto(dependencies[tableId], f.keys);
  const transitiveDependencies = doto(
    primaryDependencies,
    f.map(tblId => dependencies[tblId]),
    f.reject(f.isEmpty)
  );
  return !f.isEmpty(transitiveDependencies);
};

// This function mutates the global state object, but it shouldnt pose a problem as we replace the
// whole state anyway afterwards
export const refreshDependentRows = async (
  changeOrigin,
  changedRows,
  state
) => {
  const dependentTables = getCachedDependencyMap(state.columns);
  if (f.isEmpty(dependentTables[changeOrigin])) return state;

  const clonedState = { ...state };
  const refreshedTables = new Set();

  const fetchChangedRows = async (tableId, parentTable, changedParentRows) => {
    // assure we terminate in case of cyclic links
    if (refreshedTables.has(tableId)) {
      return;
    }
    refreshedTables.add(tableId);

    const columns = state.columns[tableId].data;
    const linkColumnIndices = doto(
      columns,
      mapIndexed((column, idx) => ({ ...column, idx })),
      f.filter(f.propEq("toTable", parentTable)),
      f.map("idx")
    );
    const getLinkCellValues = row =>
      f.flatMap(
        idx => doto(f.nth(idx, row.values), f.map(f.prop("id"))),
        linkColumnIndices
      );
    const linksToChangedRow = row =>
      doto(row, getLinkCellValues, f.any(f.contains(f.__, changedParentRows)));
    const rowsToUpdate = f
      .propOr([], ["rows", tableId, "data"], clonedState)
      .filter(linksToChangedRow);
    const fetchRows = mapPromise(({ id }) =>
      makeRequest({ apiRoute: route.toRow({ tableId, rowId: id }) })
    );
    const freshRows = await fetchRows(rowsToUpdate);
    const freshDisplayValues = freshRows.map(({ values }) =>
      mapIndexed(
        (cellValue, idx) => getDisplayValue(columns[idx], cellValue),
        values
      )
    );

    freshRows.forEach((row, ii) => {
      // Here be mutations!
      const rowIdx = f.findIndex(
        f.propEq("id", row.id),
        state.rows[tableId].data
      );
      const pathToRow = ["rows", tableId, "data", rowIdx];
      if (!f.prop(pathToRow, clonedState)) {
        set(clonedState, pathToRow, {}); // eslint-disable-line
      }
      clonedState.rows[tableId].data[rowIdx].values = row.values;
      const dvIdx = f.findIndex(
        f.propEq("id", row.id),
        state.tableView.displayValues[tableId]
      );

      const pathToDisplayValue = ["tableView", "displayValues", tableId, dvIdx];
      if (!f.prop(pathToDisplayValue, clonedState)) {
        set(clonedState, pathToDisplayValue, {}); // eslint-disable-line
      }
      clonedState.tableView.displayValues[tableId][dvIdx].values =
        freshDisplayValues[ii];
    });

    const subDependencies = dependentTables[tableId];
    return doto(
      subDependencies,
      f.keys,
      f.filter(tblId => hasTransitiveDependencies(tblId, state.columns)),
      mapPromise(tblId =>
        fetchChangedRows(tblId, tableId, rowsToUpdate.map(f.prop("id")))
      )
    );
  };

  if (hasTransitiveDependencies(changeOrigin, state.columns)) {
    await mapPromise(
      tableId => fetchChangedRows(tableId, changeOrigin, changedRows),
      f.keys(dependentTables[changeOrigin])
    );
  }
  return clonedState;
};
