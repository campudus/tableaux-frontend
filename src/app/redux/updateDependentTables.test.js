import columnResponse from "./__fixtures/modelColumns.json";

import { calcColumnDependencies } from "./updateDependentTables";

describe("Dependent state updates", () => {
  const columnCollection = { 95: { data: columnResponse.columns } };
  describe("calcColumnDependencies()", () => {
    it("does stuff", () => {
      const dependencyMap = calcColumnDependencies(columnCollection);
      expect(dependencyMap[69][95]).toEqual([60]);
    });
  });
});
