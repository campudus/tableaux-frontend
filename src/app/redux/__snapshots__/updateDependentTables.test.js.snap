// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dependent state updates calcColumnDependencies() does stuff 1`] = `
Object {
  "10": Object {
    "95": Array [
      33,
    ],
  },
  "16": Object {
    "95": Array [
      36,
    ],
  },
  "17": Object {
    "95": Array [
      37,
    ],
  },
  "18": Object {
    "95": Array [
      40,
    ],
  },
  "19": Object {
    "95": Array [
      38,
    ],
  },
  "20": Object {
    "95": Array [
      39,
    ],
  },
  "23": Object {
    "95": Array [
      14,
    ],
  },
  "25": Object {
    "95": Array [
      15,
      16,
    ],
  },
  "28": Object {
    "95": Array [
      18,
      19,
    ],
  },
  "30": Object {
    "95": Array [
      27,
      28,
    ],
  },
  "32": Object {
    "95": Array [
      20,
    ],
  },
  "33": Object {
    "95": Array [
      22,
    ],
  },
  "34": Object {
    "95": Array [
      21,
    ],
  },
  "35": Object {
    "95": Array [
      24,
    ],
  },
  "36": Object {
    "95": Array [
      23,
    ],
  },
  "37": Object {
    "95": Array [
      26,
    ],
  },
  "39": Object {
    "95": Array [
      25,
    ],
  },
  "41": Object {
    "95": Array [
      29,
    ],
  },
  "43": Object {
    "95": Array [
      30,
    ],
  },
  "44": Object {
    "95": Array [
      41,
    ],
  },
  "5": Object {
    "95": Array [
      1,
    ],
  },
  "51": Object {
    "95": Array [
      42,
    ],
  },
  "52": Object {
    "95": Array [
      43,
      44,
    ],
  },
  "54": Object {
    "95": Array [
      45,
    ],
  },
  "55": Object {
    "95": Array [
      47,
    ],
  },
  "56": Object {
    "95": Array [
      46,
    ],
  },
  "57": Object {
    "95": Array [
      50,
    ],
  },
  "59": Object {
    "95": Array [
      48,
    ],
  },
  "6": Object {
    "95": Array [
      32,
    ],
  },
  "60": Object {
    "95": Array [
      49,
    ],
  },
  "62": Object {
    "95": Array [
      52,
    ],
  },
  "64": Object {
    "95": Array [
      51,
    ],
  },
  "65": Object {
    "95": Array [
      55,
    ],
  },
  "66": Object {
    "95": Array [
      54,
    ],
  },
  "67": Object {
    "95": Array [
      53,
    ],
  },
  "69": Object {
    "95": Array [
      60,
    ],
  },
  "7": Object {
    "95": Array [
      31,
    ],
  },
  "70": Object {
    "95": Array [
      34,
    ],
  },
  "71": Object {
    "95": Array [
      61,
    ],
  },
  "72": Object {
    "95": Array [
      59,
    ],
  },
  "74": Object {
    "95": Array [
      58,
    ],
  },
  "75": Object {
    "95": Array [
      57,
    ],
  },
  "77": Object {
    "95": Array [
      56,
    ],
  },
  "78": Object {
    "95": Array [
      6,
    ],
  },
  "79": Object {
    "95": Array [
      7,
    ],
  },
  "80": Object {
    "95": Array [
      8,
    ],
  },
  "84": Object {
    "95": Array [
      3,
    ],
  },
  "85": Object {
    "95": Array [
      62,
    ],
  },
  "92": Object {
    "95": Array [
      35,
    ],
  },
  "94": Object {
    "95": Array [
      4,
    ],
  },
}
`;
